Switch cookie accept policy when using managed domains

brentfulgham opened the pull request:
#### 8d0a1a99ffad4617e22c1ee48442d755a11488ae
<pre>
Switch cookie accept policy when using managed domains
<a href="https://bugs.webkit.org/show_bug.cgi?id=262415">https://bugs.webkit.org/show_bug.cgi?id=262415</a>
&lt;rdar://116025019&gt;

Reviewed by Pascoe.

The ITP features block third party cookie access, and purge local state regularly.
When an MDM application is running with managed domains, we should adjust cookie
acceptance policy to support those managed origins.

* Source/WebKit/NetworkProcess/NetworkProcess.cpp:
(WebKit::NetworkProcess::setManagedDomainsForResourceLoadStatistics):

Canonical link: <a href="https://commits.webkit.org/268871@main">https://commits.webkit.org/268871@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/66211b7bea5c4a71e3320a5813e8b65ac129962c

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/20777 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/21185 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/21844 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/22667 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/19376 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/24429 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/21365 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/20686 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/20998 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/15/builds/20787 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/18046 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/23520 "Built successfully") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/39/builds/17953 "Passed tests") | | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/25167 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/19031 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/19033 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/23067 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/19626 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/17/builds/16676 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/18857 "Built successfully") | | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/5019 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/23188 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/19444 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

pascoej commented on Source/WebKit/NetworkProcess/NetworkProcess.cpp:
```diff
@@ -1330,10 +1330,16 @@ void NetworkProcess::setAppBoundDomainsForResourceLoadStatistics(PAL::SessionID
-            resourceLoadStatistics->setManagedDomains(WTFMove(appBoundDomains), WTFMove(completionHandler));
-            return;
-        }
+        supplement<WebCookieManager>()->setHTTPCookieAcceptPolicy(sessionID, WebCore::HTTPCookieAcceptPolicy::AlwaysAccept, [session = WeakPtr { *session }, appBoundDomains = WTFMove(appBoundDomains), completionHandler = WTFMove(completionHandler)]() mutable {
```
nit (especially because I introduced it): 
```suggestion
        supplement<WebCookieManager>()->setHTTPCookieAcceptPolicy(sessionID, WebCore::HTTPCookieAcceptPolicy::AlwaysAccept, [session = WeakPtr { *session }, managedDomains = WTFMove(managedDomains), completionHandler = WTFMove(completionHandler)]() mutable {
```

brentfulgham replied:
Oh gosh! I didn't even notice! LOL.

brentfulgham replied:
Will fix.

szewai reviewed (commented):
Why do we only need to switch CookieAcceptPolicy to Always when managed domains are set? Should we switch it when `NetworkSession::m_thirdPartyCookieBlockingMode` is not `ThirdPartyCookieBlockingMode::All`?

sysrqb commented on Source/WebKit/NetworkProcess/NetworkProcess.cpp:
```diff
@@ -1327,13 +1327,19 @@ void NetworkProcess::setAppBoundDomainsForResourceLoadStatistics(PAL::SessionID
-        }
+        supplement<WebCookieManager>()->setHTTPCookieAcceptPolicy(sessionID, WebCore::HTTPCookieAcceptPolicy::AlwaysAccept, [session = WeakPtr { *session }, managedDomains = WTFMove(managedDomains), completionHandler = WTFMove(completionHandler)]() mutable {
+            if (session) {
+                if (auto* resourceLoadStatistics = session->resourceLoadStatistics()) {
```
Without making this too complicated, you might want to think about the correct behavior if this condition fails - should we still allow all cookies in this case?

brentfulgham replied:
That's an interesting point. If we have reached this point, the MDM administrator told WebKit it wanted to operate in this special mode, but for some reason the `NetworkSession` was killed between the time we told CFNetwork about it, and the time we went to update the ITP infrastructure about the domains.

If the `NetworkSession` associated with the `sessionID` is dead, we can't invoke `setHTTPCookieAcceptPolicy` again to change it, because we have no `sessionID` to give it. But that's probably okay, since the cookie state is set on the `CFHTTPCookieStorageRef` associated with `sessionID`. If the session no longer exists, the cookie store should also be gone (at least for Cocoa-based ports).

So I think we either want to leave the code as-is (since the re-spawned NetworkSession will pass through this code again and set the HTTP Cookie Accept policy to the same way, so there would be no behavior change), OR we should RELEASE_ASSERT that session is not `nullptr` and just crash so the UIProcess respawns if a NetworkSession

brentfulgham committed:
Switch cookie accept policy when using managed domains
https://bugs.webkit.org/show_bug.cgi?id=262415
<rdar://116025019>

Reviewed by Pascoe.

The ITP features block third party cookie access, and purge local state regularly.
When an MDM application is running with managed domains, we should adjust cookie
acceptance policy to support those managed origins.

* Source/WebKit/NetworkProcess/NetworkProcess.cpp:
(WebKit::NetworkProcess::setManagedDomainsForResourceLoadStatistics):

Canonical link: https://commits.webkit.org/268871@main

johnwilander reviewed (commented):
Are we running any of our cookie blocking tests under this setting? We should make sure that we haven't regressed the cookie blocking that should remain. Relaxing cookie blocking is one of those things that typically just passes more tests unless you are testing the blocking.
